dig.go

```diff
@@ -0,0 +1,281 @@
+//
+// Any function passed to Provide is assumed to be a constructor. Constructors
+// can take any number of parameters, which will be supplied by the Container
+// on demand. They must return at least one non-error value, all of which are
+// then available in the Container. If the last returned value is an error, the
+// Container inspects it to determine whether the constructor succeeded or
+// failed. Regardless of position, returned errors are never put into the
+// Container's dependency graph.
+//
+// All non-functions (including structs, pointers, Go's built-in collections,
+// and primitive types like ints) are inserted into the Container as-is.
+func (c *Container) Provide(constructor interface{}) error {
+	ctype := reflect.TypeOf(constructor)
+	if ctype == nil {
+		return errors.New("can't provide an untyped nil")
+	}
+	// Since we want to wrap any errors, don't return early.
+	var err error
+	if ctype.Kind() != reflect.Func {
+		err = c.provideInstance(constructor)
+	} else {
+		err = c.provideConstructor(constructor, ctype)
+	}
+
+	if err == nil {
+		return nil
+	}
+	return fmt.Errorf("can't provide %v: %v", ctype, err)
+}
+
+// Invoke runs a function, supplying its arguments from the Container. If the
+// function's last return value is an error, that error is propagated to the
+// caller. All other returned values (if any) are ignored.
+//
+// Passing anything other than a function to Invoke returns an error
+// immediately.
+func (c *Container) Invoke(function interface{}) error {
+	ftype := reflect.TypeOf(function)
+	if ftype == nil {
+		return errors.New("can't invoke an untyped nil")
+	}
+	if ftype.Kind() != reflect.Func {
+		return fmt.Errorf("can't invoke non-function %v (type %v)", function, ftype)
+	}
+	args, err := c.constructorArgs(ftype)
+	if err != nil {
+		return fmt.Errorf("failed to get arguments for %v (type %v): %v", function, ftype, err)
+	}
+	returned := reflect.ValueOf(function).Call(args)
+	if len(returned) == 0 {
+		return nil
+	}
+	if last := returned[len(returned)-1]; last.Type() == _errType && last.Interface() != nil {
+		return fmt.Errorf("failed to execute %v (type %v): %v", function, ftype, last.Interface())
+	}
+	return nil
+}
+
+func (c *Container) provideInstance(val interface{}) error {
+	vtype := reflect.TypeOf(val)
+	if vtype == _errType {
+		return errors.New("can't provide errors")
+	}
+	if _, ok := c.nodes[vtype]; ok {
+		return errors.New("already in container")
+	}
+	c.nodes[vtype] = node{provides: vtype}
+	c.cache[vtype] = reflect.ValueOf(val)
+	return nil
+}
+
+func (c *Container) provideConstructor(ctor interface{}, ctype reflect.Type) error {
+	returnTypes := make(map[reflect.Type]struct{}, ctype.NumOut())
+	for i := 0; i < ctype.NumOut(); i++ {
+		rt := ctype.Out(i)
+		if rt == _errType {
+			// Don't register errors into the container.
+			continue
+		}
+		if _, ok := returnTypes[rt]; ok {
+			return fmt.Errorf("returns multiple %v", rt)
```

Comment:
Yes. I don't see a ton of value in annotating and capturing stacktraces in dig, since it's easy enough to do with `fmt.Errorf`.